RecoLocalTracker/SiPhase2VectorHitBuilder/test/ClustersValidationTGraph.cc

```diff
@@ -0,0 +1,618 @@
+#include <map>
+#include <vector>
+#include <algorithm>
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+#include "Geometry/TrackerGeometryBuilder/interface/PixelGeomDetUnit.h"
+
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/Common/interface/DetSetVectorNew.h"
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Phase2TrackerCluster/interface/Phase2TrackerCluster1D.h"
+#include "DataFormats/Phase2TrackerDigi/interface/Phase2TrackerDigi.h"
+#include "DataFormats/SiPixelDigi/interface/PixelDigi.h"
+
+#include "SimDataFormats/TrackerDigiSimLink/interface/PixelDigiSimLink.h"
+#include "SimDataFormats/Track/interface/SimTrackContainer.h"
+#include "SimDataFormats/Vertex/interface/SimVertexContainer.h"
+#include "SimDataFormats/TrackingHit/interface/PSimHitContainer.h"
+
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+#include "CommonTools/Utils/interface/TFileDirectory.h"
+
+#include <TGraph.h>
+#include <TH1F.h>
+#include <THStack.h>
+#include <TTree.h>
+
+struct ClusterHistos {
+  THStack* numberClustersMixed;
+  TH1F* numberClusterPixel;
+  TH1F* numberClusterStrip;
+
+  THStack* clustersSizeMixed;
+  TH1F* clusterSizePixel;
+  TH1F* clusterSizeStrip;
+
+  TGraph* globalPosXY[3];
+  TGraph* localPosXY[3];
+
+  TH1F* deltaXClusterSimHits[3];
+  TH1F* deltaYClusterSimHits[3];
+
+  TH1F* deltaXClusterSimHits_P[3];
+  TH1F* deltaYClusterSimHits_P[3];
+
+  TH1F* digiEfficiency[3];
+
+  TH1F* primarySimHits;
+  TH1F* otherSimHits;
+};
+
+class Phase2TrackerClusterizerValidationTGraph : public edm::EDAnalyzer {
```

Comment:
should this code be moved to a Validation package?

Reply:
This comment applies to all the .cc files introduced in this folder (which also need a lot of cleanup)

Reply:
Yes, this stuff will be moved to a validation package soon-ish.

Reply:
Yes, sorry for that.
But it is not the only case:
- https://github.com/cms-sw/cmssw/tree/master/RecoLocalTracker/SiPhase2Clusterizer/test
- https://github.com/cms-sw/cmssw/tree/master/RecoLocalTracker/SiPixelClusterizer/test
- https://github.com/cms-sw/cmssw/tree/master/RecoLocalTracker/SiStripRecHitConverter/test
...
Is there a rule or speacial reasons of why this is not implemented in many cases in `RecoLocalTracker`?

Reply:
It is probably a historical remnant, but it would be good to have better organization in new code.

Reply:
Of course. Would it be ok to leave it as is for now and move it later when we integrate the VH validation into the full tracking validation setup?

Reply:
If it is really necessary to have this code in this PR and this location, yes, it is acceptable. (As always, the concern is that the probability of a followup item actually being completed is inversely proportional to how much work it takes.)